File: leetcode/src/00279_perfect_squares/solution.go
package leetcode

func numSquares(n int) int {
	visited := map[int]struct{}{}
	q := []int{0}
	visited[0] = struct{}{}
	res := 0
	for len(q) != 0 {
		for i := len(q); i > 0; i-- {
			var front int
			front, q = q[0], q[1:]
			if front == n {
				return res
			}
			for i := 1; i*i <= n-front; i++ {
				newVal := front + i*i
				if _, ok := visited[newVal]; ok {
					continue
				}
				visited[newVal] = struct{}{}
				q = append(q, newVal)
			}
		}
		res++
	}
	return res
}

File: leetcode/src/00279_perfect_squares/solution.rs
struct Solution;

impl Solution {
    pub fn num_squares(n: i32) -> i32 {
        let n = n as usize;
        let arr = Solution::square_numbers(n);
        let mut dp = vec![n + 1; n + 1];
        dp[0] = 0;
        for &val in &arr {
            for i in val..=n {
                dp[i] = dp[i].min(dp[i - val] + 1);
            }
        }
        dp[n] as _
    }

    fn square_numbers(limit: usize) -> Vec<usize> {
        let mut res = vec![];
        let mut i = 1;
        while i * i <= limit {
            res.push(i * i);
            i += 1;
        }
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let n = 12;
        assert_eq!(Solution::num_squares(n), 3);

        let n = 13;
        assert_eq!(Solution::num_squares(n), 2);
    }
}
